protocol/x/vault/keeper/orders.go

```diff
@@ -72,78 +73,46 @@ func (k Keeper) RefreshAllVaultOrders(ctx sdk.Context) {
-			}
-			vaultId.IncrCounterWithLabels(
-				metrics.VaultCancelOrder,
-				metrics.GetLabelForBoolValue(metrics.Success, err == nil),
-			)
-		}
-	}
-	// Place new CLOB orders.
+	// Get client IDs of most recently placed orders, if any.
+	mostRecentClientIds := k.GetMostRecentClientIds(ctx, vaultId)
+	// Get orders to place.
 	ordersToPlace, err := k.GetVaultClobOrders(ctx, vaultId)
 	if err != nil {
 		log.ErrorLogWithError(ctx, "Failed to get vault clob orders to place", err, "vaultId", vaultId)
 		return err
 	}
 
-	for i, order := range ordersToPlace {
-		err := k.PlaceVaultClobOrder(ctx, order)
-		if err != nil {
-			log.ErrorLogWithError(ctx, "Failed to place order", err, "order", order, "vaultId", vaultId)
-		}
-
-		vaultId.IncrCounterWithLabels(
-			metrics.VaultPlaceOrder,
-			metrics.GetLabelForBoolValue(metrics.Success, err == nil),
-		)
-
-		// Send indexer messages. We expect orderIdsToCancel and ordersToPlace to have the same length
-		// and the order to place at each index to be a replacement of the order to cancel at the same index.
-		replacedOrderId := orderIdsToCancel[i]
-		if replacedOrderId == nil {
-			k.GetIndexerEventManager().AddTxnEvent(
-				ctx,
-				indexerevents.SubtypeStatefulOrder,
-				indexerevents.StatefulOrderEventVersion,
-				indexer_manager.GetBytes(
-					indexerevents.NewLongTermOrderPlacementEvent(
-						*order,
-					),
-				),
-			)
+	clientIds := make([]uint32, len(ordersToPlace))
+	for i, orderToPlace := range ordersToPlace {
+		if i >= len(mostRecentClientIds) { // when a vault first starts quoting or when `layers` increases.
+			// Place order.
+			err = k.PlaceVaultClobOrder(ctx, vaultId, orderToPlace)
 		} else {
-			k.GetIndexerEventManager().AddTxnEvent(
-				ctx,
-				indexerevents.SubtypeStatefulOrder,
-				indexerevents.StatefulOrderEventVersion,
-				indexer_manager.GetBytes(
-					indexerevents.NewLongTermOrderReplacementEvent(
-						*replacedOrderId,
-						*order,
-					),
-				),
-			)
+			oldClientId := mostRecentClientIds[i]
+			oldOrderId := vaultId.GetClobOrderId(oldClientId)
+			oldOrderPlacement, exists := k.clobKeeper.GetLongTermOrderPlacement(ctx, *oldOrderId)
+			if !exists {
+				// Place order.
+				err = k.PlaceVaultClobOrder(ctx, vaultId, orderToPlace)
+			} else if oldOrderPlacement.Order.Quantums != orderToPlace.Quantums ||
+				oldOrderPlacement.Order.Subticks != orderToPlace.Subticks {
+				// Replace old order with new order.
+				// Flip last bit of old client ID to get new client ID to make sure they are different
+				// as order placement fails if the same order ID is already marked for cancellation.
+				orderToPlace.OrderId.ClientId = oldClientId ^ 1
+				err = k.ReplaceVaultClobOrder(ctx, vaultId, oldOrderId, orderToPlace)
+			}
+		}
+		if err != nil {
+			log.ErrorLogWithError(ctx, "Failed to place/replace vault clob order", err, "vaultId", vaultId)
 		}
+		clientIds[i] = orderToPlace.OrderId.ClientId
```

Comment:
nit: Make it explicit that if the order is not replaced `oldClientId` is what' stored into the `clientIds`.